context/app/static/js/shared-styles/dropdowns/DropdownListbox/DropdownListbox.jsx

```diff
@@ -0,0 +1,83 @@
+import React, { useState, useRef } from 'react';
+import PropTypes from 'prop-types';
+import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDownRounded';
+import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUpRounded';
+import ClickAwayListener from '@material-ui/core/ClickAwayListener';
+import MenuList from '@material-ui/core/MenuList';
+
+import { StyledPopper, StyledPaper } from './style';
+
+function DropdownListbox(props) {
+  const {
+    selectedOptionIndex,
+    buttonComponent: SelectionButton,
+    optionComponent: Option,
+    buttonProps,
+    options,
+    selectOnClick,
+    getOptionLabel,
+    id,
+  } = props;
+  const anchorRef = useRef(null);
+  const [isOpen, setIsOpen] = useState(false);
+
+  function selectOption(option) {
+    selectOnClick(option);
+    setIsOpen(false);
+  }
+
+  return (
+    <>
+      <SelectionButton
+        ref={anchorRef}
+        id={`${id}-button`}
+        aria-haspopup="listbox"
+        disableElevation
+        variant="contained"
+        color="primary"
+        onClick={() => setIsOpen(true)}
+        {...buttonProps}
+      >
+        {getOptionLabel(options[selectedOptionIndex])} {isOpen ? <ArrowDropUpIcon /> : <ArrowDropDownIcon />}
+      </SelectionButton>
+      <StyledPopper open={isOpen} anchorEl={anchorRef.current} placement="bottom-start">
+        <StyledPaper>
+          <ClickAwayListener onClickAway={() => setIsOpen(false)}>
+            <MenuList role="listbox" id={`${id}-options`}>
+              {options.map((option, i) => (
+                <Option
+                  onClick={() => selectOption({ option, i })}
+                  key={getOptionLabel(option)}
+                  autoFocus={selectedOptionIndex === i}
+                  selected={selectedOptionIndex === i}
+                  aria-selected={selectedOptionIndex === i}
+                >
+                  {getOptionLabel(option)}
+                </Option>
+              ))}
+            </MenuList>
+          </ClickAwayListener>
+        </StyledPaper>
+      </StyledPopper>
+    </>
+  );
+}
+
+DropdownListbox.propTypes = {
+  selectedOptionIndex: PropTypes.number.isRequired,
+  buttonComponent: PropTypes.elementType.isRequired,
+  optionComponent: PropTypes.elementType.isRequired,
+  /* eslint-disable react/forbid-prop-types */
+  buttonProps: PropTypes.object,
+  options: PropTypes.array.isRequired,
+  /* eslint-enable react/forbid-prop-types */
```

Comment:
How do you feel about these eslint rules overall? Maybe it's fine: it still catches someone passing entirely the wrong thing... or is it good to have something nudge us to complete it more fully?

Reply:
I think it does make sense to be more specific where possible and eslint nudges us in that direction, but in situations like this where it can be an array of many different types it is simplest to just ignore it. I'm open to doing something different if you have a stronger preference.

Reply:
Fine as-is, just wanted to check. Thanks!